runtime/domain.c

```diff
@@ -630,28 +699,146 @@ CAMLexport void caml_reset_domain_lock(void)
   return;
 }
 
-void caml_init_domains(uintnat minor_heap_wsz) {
-  int i;
-  uintnat size;
-  void* heaps_base;
+/* minor heap initialization and resizing */
 
+static uintnat get_minor_heap_reservation_size() {
   /* sanity check configuration */
-  if (caml_mem_round_up_pages(Bsize_wsize(Minor_heap_max))
-          != Bsize_wsize(Minor_heap_max))
-    caml_fatal_error("Minor_heap_max misconfigured for this platform");
+  if (caml_mem_round_up_pages(Bsize_wsize(caml_minor_heap_max_wsz))
+          != Bsize_wsize(caml_minor_heap_max_wsz))
+    caml_fatal_error("minor_heap_max misconfigured for this platform");
+
+  return ((uintnat)Bsize_wsize(caml_minor_heap_max_wsz)
+          * Max_domains);
+}
+
+static void reserve_minor_heaps(uintnat size) {
+  void* heaps_base;
 
   /* reserve memory space for minor heaps */
-  size = (uintnat)Bsize_wsize(Minor_heap_max) * Max_domains;
   heaps_base = caml_mem_map(size, caml_sys_pagesize, 1 /* reserve_only */);
   if (heaps_base == NULL)
-    caml_fatal_error("Not enough heap memory to start up");
+    caml_fatal_error("Not enough heap memory to reserve minor heaps");
 
   caml_minor_heaps_base = (uintnat) heaps_base;
   caml_minor_heaps_end = (uintnat) heaps_base + size;
 
+  caml_gc_log("new minor heap reserved from %p to %p",
+              (value*)caml_minor_heaps_base, (value*)caml_minor_heaps_end);
+
+  for (int i = 0; i < Max_domains; i++) {
+    struct dom_internal* dom = &all_domains[i];
+
+    uintnat domain_minor_heap_area = caml_minor_heaps_base +
+      (uintnat)Bsize_wsize(caml_minor_heap_max_wsz) * (uintnat)i;
+
+    dom->minor_heap_area = domain_minor_heap_area;
+    dom->minor_heap_area_end =
+         domain_minor_heap_area + Bsize_wsize(caml_minor_heap_max_wsz);
+
+    CAMLassert(dom->minor_heap_area_end <= caml_minor_heaps_end);
+  }
+}
+
+static void unreserve_minor_heaps() {
+  uintnat size;
+
+  caml_gc_log("unreserve_minor_heaps");
+
+  for (int i = 0; i < Max_domains; i++) {
+    struct dom_internal* dom = &all_domains[i];
+
+    CAMLassert(
+      /* this domain is not running */
+      !dom->interruptor.running
+      || (
+        /* or its minor heap must already be uninitialized */
+        dom->state != NULL
+        && dom->state->young_start == NULL
+        && dom->state->young_end == NULL
+      ));
+    /* Note: interruptor.running does not guarantee that dom->state is
+       correctly initialized, but domain initialization cannot run
+       concurrently with STW sections so we cannot observe partial
+       initialization states. */
+
+    /* uninitialize the minor heap area */
+    dom->minor_heap_area = dom->minor_heap_area_end = 0;
+  }
+
+  size = caml_minor_heaps_end - caml_minor_heaps_base;
```

Comment:
I assume
```c
CAMLassert (Bsize_wsize(caml_minor_heap_max_wsz) * Max_domains == size)
```
holds here.

Reply:
I think so! In the past we would update `caml_minor_heap_wsz` to the new/desired value *before* calling this function, so we needed to be careful to use the "real size" and not the "desired size" here. But the current version is more robust (no slightly inconsistent state), and I think the proposed assert holds. (It could be falsified again with more flexible heap-area-reservation schemes, for example if we used an exponential-resizing strategy instead of requesting exactly the new needed size each time.)

Reply:
Would the assertion still be useful to document the current behaviour?

Reply:
Yes of course; I am planning to take a global look at all comments and update the PR. (But probably later in the day or week.)

Reply:
I added the assert.